### PR TITLE
Fix iteration for RocksBlockStore

### PR DESCRIPTION
**Pull request by ggezer**
This is to get rocks db out of prefix seek mode, under which iterators seekToFirst/Last APIs are ineffective. See https://github.com/facebook/rocksdb/wiki/Prefix-Seek-API-Changes for more context.

Fixes https://github.com/Alluxio/alluxio/issues/10858

### CODE REVIEW

**Comment by ZacBlanco** [comment]
@ggezer can you update the description to describe how this PR fixes the issue?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 1]
```diff
@@ -1516,6 +1516,13 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.ENFORCE)
           .setScope(Scope.MASTER)
           .build();
+  public static final PropertyKey MASTER_METASTORE_BLOCK_ITERATOR_READAHEAD_SIZE =
+      new Builder(Name.MASTER_METASTORE_BLOCK_ITERATOR_READAHEAD_SIZE)
+          .setDefaultValue("10000")
+          .setDescription("The read-ahead size for block iterator.")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]: Can you describe this more, and the implications of a small one and a large one?

Also, what is "size"? rows? bytes?

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java [hunk 2]
```diff
@@ -67,10 +70,11 @@
   public RocksBlockStore(String baseDir) {
     RocksDB.loadLibrary();
     mDisableWAL = new WriteOptions().setDisableWAL(true);
+    mIteratorOption = new ReadOptions().setReadaheadSize(
+        ServerConfiguration.getInt(PropertyKey.MASTER_METASTORE_BLOCK_ITERATOR_READAHEAD_SIZE));
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java]: below, it uses `setPrefixSameAsStart()`. We don't need that anymore?

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java [hunk 3]
```diff
@@ -67,10 +70,11 @@
...
     ColumnFamilyOptions cfOpts = new ColumnFamilyOptions()
         .setMemTableConfig(new HashLinkedListMemTableConfig())
-        .setCompressionType(CompressionType.NO_COMPRESSION)
-        .useFixedLengthPrefixExtractor(8); // We always search using the initial long key
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java]: why is this removed?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 4]
```diff
@@ -1516,6 +1516,13 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.ENFORCE)
           .setScope(Scope.MASTER)
           .build();
+  public static final PropertyKey MASTER_METASTORE_BLOCK_ITERATOR_READAHEAD_SIZE =
+      new Builder(Name.MASTER_METASTORE_BLOCK_ITERATOR_READAHEAD_SIZE)
+          .setDefaultValue("10000")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]: Should we make this larger?

File: core/server/master/src/test/java/alluxio/master/metastore/rocks/RocksBlockStoreTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  public void testIterator() throws Exception {
+    final int blockCount = 3;
+    RocksBlockStore blockStore = new RocksBlockStore(mFolder.newFolder().getAbsolutePath());
+    for (int i = 0; i < blockCount; i++) {
+      blockStore.putBlock(i, Block.BlockMeta.newBuilder().setLength(i).build());
+    }
+
+    Iterator<BlockStore.Block> iter = blockStore.iterator();
+    for (int i = 0; i < blockCount; i++) {
+      assertTrue(iter.hasNext());
```
Comment (gpang) [inline core/server/master/src/test/java/alluxio/master/metastore/rocks/RocksBlockStoreTest.java]: after the loop, we should check that the iterator is done?

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java (hunk 3 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java]: This will put rocks-db to prefix-seek mode. Please see description for more details.

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 1 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: bytes

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 4 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: I was under impression that it was count, but turns out it's bytes. Made it 4MB to start with.

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java (hunk 2 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockStore.java]: This one uses `RocksIterator.seek()` for which this is fine.

**Comment by ggezer** [comment]
jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java [hunk 6]
```diff
@@ -245,7 +245,8 @@ public boolean hasChildren(InodeDirectoryView inode, ReadOption option) {
    * @return an iterator over stored inodes
    */
   public Iterator<InodeView> iterator() {
-    return new RocksInodeViewIterator(db().newIterator(mInodesColumn.get()));
+    return RocksUtils.createIterator(db().newIterator(mInodesColumn.get()),
+        (iter) -> getMutable(Longs.fromByteArray(iter.key()), ReadOption.defaults()).get());
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksInodeStore.java]: should this also use the read option like the other iterator?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 1 above)
Comment (gpang) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: I see. Could you update the description to mention bytes, and maybe some of the implications of the size? For example, if we have a billion blocks, is reading 4mb at a time efficient enough? Should this be a larger size?

Comment (ggezer) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: I guess 4MB suffices, as it'd prefetch about 10K inodes and 100K block-Ids. This should be enough as iterator content are ultimately written back to disk.

**Comment by ggezer** [comment]
alluxio-bot, merge this please
